timemachine/fe/interaction_group_traj.py

```diff
@@ -0,0 +1,193 @@
+
+        # note: vmap here can consume excessive memory if len(env_idxs) * len(xs) is large -> python loop
+        # mask = vmap(env_mask_within_cutoff, (0,0,0,None))(_xs_env, self.xs_lig, boxes, cutoff)
+        def f(x_env, x_lig, box):
+            return env_mask_within_cutoff(x_env, x_lig, box, cutoff)
+
+        mask = np.array([f(_xs_env[i], self.xs_lig[i], np.diag(box_diags[i])) for i in range(self.n_frames)])
+
+        padded_num_env_atoms = mask.sum(1).max()
+
+        if verbose:
+            num_stored = padded_num_env_atoms + len(ligand_idxs)
+            max_nbrs, mean_nbrs = padded_num_env_atoms, mask.sum(1).mean()
+            msg = f"""
+                saving {(xs.shape[1] / num_stored):.2f}x on storage
+                (relative to storing all env atoms)
+                padding to max_nbrs = {max_nbrs}
+                (~{max_nbrs/mean_nbrs:.2f}x larger than unpadded neighbor list)
+            """
+            print(msg)
+
+        idxs_within_env_block = np.argsort(mask, axis=1)[:, -padded_num_env_atoms:]
+        self.selected_env_idxs = jnp.array(all_env_idxs[idxs_within_env_block], dtype=jnp.uint32)
+
+        self.xs_env = np.array([_x_env[idxs] for (_x_env, idxs) in zip(_xs_env, idxs_within_env_block)])
+        self.box_diags = box_diags
+
+    def to_dict(self):
+        return dict(
+            xs_lig=np.array(self.xs_lig),
+            xs_env=np.array(self.xs_env),
+            box_diags=np.array(self.box_diags),
+            cutoff=self.cutoff,
+            selected_env_idxs=np.array(self.selected_env_idxs),
+            ligand_idxs=np.array(self.ligand_idxs),
+        )
+
+    @classmethod
+    def from_dict(cls, dict_of_np_arrays):
+        archive = dict_of_np_arrays
+        traj = cls.__new__(cls)
+        traj.xs_lig = archive["xs_lig"]
+        traj.xs_env = archive["xs_env"]
+        traj.box_diags = archive["box_diags"]
+        traj.cutoff = archive["cutoff"]
+        traj.selected_env_idxs = archive["selected_env_idxs"]
+        traj.ligand_idxs = archive["ligand_idxs"]
+        traj.n_frames = len(traj.xs_env)
+        return traj
+
+    def to_npz(self, fname):
+        np.savez_compressed(fname, **self.to_dict())
+
+    @classmethod
+    def from_npz(cls, fname):
+        npz_archive = np.load(fname, allow_pickle=False)
+        return cls.from_dict(npz_archive)
+
+    def make_U_fxn(self, pair_fxn: PairFxn):
+        """
+
+        Parameters
+        ----------
+        pair_fxn : pair_fxn(x_i, x_j, param_i, param_j, box) -> energy
+            scalar-valued jax-transformable function of a pair of particles and their parameters
+
+        Returns
+        -------
+        vector-valued function of nb params
+        """
+
+        # apply pair_fxn(x_i, x_j, param_i, param_j, box)
+        axes_a = (0, None, 0, None, None)
+        axes_b = (None, 0, None, 0, None)
+        all_pairs_fxn = vmap(vmap(pair_fxn, axes_a), axes_b)
```

Comment:
Is vmap here particularly expensive? i.e. why not just do this in compute_Us(pair_fxn, nb_params)?

Reply:
Not expensive, just style preference (define once, invoke >once).

Since this definition doesn't depend on `params`, this seems like a natural place for this definition, and it keeps a few lines of clutter out of the body of `compute_Us(params)`.